### PR TITLE
Socket Mode Rewrite to Python(ish) Implementation

### PR DESCRIPTION
**Pull request by filmaj**
# Socket Mode Rewrite

This PR ports, as much as possible, the Python implementation of the Socket Mode client to node.

I have been running a longevity test of an app, in production, on a busy workspace, using this PR since April 24th without issue (the app reconnects gracefully when Slack sends a `disconnect` message, which is what we expect).

## Why?

The existing node socket-mode package uses `finity`, an old finite state machine configuration package. While it has served us well, it has been difficult to maintain, with tricky semantics. Additionally, the existing socket-mode implementation is almost a straight copy-paste of the `rtm-api` module, which is very old.

## Breaking Changes

- Removed two [lifecycle events](https://github.com/slackapi/node-slack-sdk/tree/main/packages/socket-mode#lifecycle-events) - though we can add them back in if we think they are useful / worth keeping for backwards compatibility:
  - `authenticated`. I did not find it useful, it just represented that the client retrieved the WS URL to connect to via a successful response from the `apps.connections.open` HTTP API.
  - `unable_to_start_socket_mode`. I felt like the `error` event was sufficient?
- Some properties and methods that were previously `public` on `SocketModeClient` have been removed: `connected`, `authenticated` and `isActive()`, though `isActive()` could easily be added back in.

## What's The Same?

Many of the developer-facing APIs have remained the same:

- the `EventEmitter`-based design, where the client emits the events described in the README, remain identical. This should ease the upgrading path for any projects dependent on this package, including bolt-js.
- I think `SocketModeClient`'s standard public API also remains the same? `start()` and `disconnect()` for sure, but `isActive()` is now moved to `SlackWebSocket`, so that's a major change - though to be fair, we never documented these APIs.

## Internal Design Changes

- `SocketModeClient` aims to implement similar interfaces as [Python's `SocketModeClient` class, as well as its `BaseSocketModeClient`](https://github.com/slackapi/python-slack-sdk/blob/main/slack_sdk/socket_mode/builtin/client.py).
  - The `Client` is responsible for retrieving the WebSocket URL to connect to (by interacting with the [`apps.connections.open` HTTP API](https://api.slack.com/methods/apps.connections.open)) and managing reconnections. We can consider these as _application-level_ concerns.
- `SlackWebSocket` aims to implement similar interfaces as [Python's `Connection` class](https://github.com/slackapi/python-slack-sdk/blob/main/slack_sdk/socket_mode/builtin/connection.py). Luckily for the node implementation, many of the low-level details of the WebSocket protocol are encapsulated in the `ws` dependency.
  - `SlackWebSocket` concerns itself with managing and maintaining a single WebSocket connection and nothing more; any WS-specific events relevant to application functioning (e.g. disconnects) are raised back to the `Client` for it to deal with.

IMO this leads to a nice separation of concerns.

## Concerns / Questions / Request for Feedback

- Moving to an object-oriented / class-based setup, actually unit testing the private methods in JavaScript is.. impossible? So I think the code coverage from unit tests will drop significantly as a result. Any suggestions on how to tackle this? Could just make methods public and prefix with an `_` just to expose them for unit tests?
- Seems like the documentation on https://slack.dev/node-slack-sdk/socket-mode is just a copy paste of the README (and even the comments in the readme markdown state as much). Can we instead link to the README from the slack.dev site? I feel like the chances of the package documentation being kept up to date are better by being in the README.

### CODE REVIEW

File: docs/_packages/socket_mode.md [hunk 1]
```diff
@@ -7,133 +7,11 @@ anchor_links_header: Usage
 
 # Slack Socket Mode
 
+For baseline package documentation, please see the project's [`README`](https://github.com/slackapi/node-slack-sdk/tree/main/packages/socket-mode#readme) or the [documentation on npm](https://www.npmjs.com/package/@slack/socket-mode).
```
Comment (filmaj) [inline docs/_packages/socket_mode.md:10]: Mainly axing content in here, as it used to be a straight copy/paste from the socket mode README. No need to duplicate, just link to the duplicated content and keep this area for additional documentation (if necessary).

File: packages/socket-mode/package.json [hunk 2]
```diff
@@ -18,8 +18,8 @@
     "state",
     "connection"
   ],
-  "main": "dist/index.js",
-  "types": "./dist/index.d.ts",
+  "main": "dist/src/index.js",
+  "types": "./dist/src/index.d.ts",
```
Comment (filmaj) [inline packages/socket-mode/package.json:22]: These changes are needed to enable the `resolveJsonModule` option in the `tsconfig.json`, as setting that option changes the output of the TypeScript compiler.

File: packages/socket-mode/src/logger.ts [hunk 3]
```diff
@@ -1,26 +1,25 @@
...
 
-/**
- * INTERNAL interface for getting or creating a named Logger.
- */
-export function getLogger(name: string, level: LogLevel, existingLogger?: Logger): Logger {
-  // Get a unique ID for the logger.
-  const instanceId = instanceCount;
-  instanceCount += 1;
+export default {
+  getLogger: function getLogger(name: string, level: LogLevel, existingLogger?: Logger): Logger {
```
Comment (filmaj) [inline packages/socket-mode/src/logger.ts:8]: This restructuring allows for easier mocking without the need for a tool like `rewire` or `proxyquire`.

**Review by WilliamBergamin — COMMENTED** [review]
This is awesome 💯 I'm a fan of the OOP approach, this version is definitely easier to understand 

I think respecting object encapsulation instead of making things public with `_` prefix will yield more maintainable code
Faking/Mocking a socket mode server to test the the intended behavior of `SlackWebSocket` seems like a viable long term testing solution since this API is stable

File: packages/socket-mode/src/SlackWebSocket.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+  /**
+   * Returns true if the underlying WebSocket connection is active.
+   */
+  public isActive(): boolean { // python equiv: SocketModeClient.is_connected
+    if (!this.websocket) {
+      this.logger.debug('isActive(): websocket not instantiated!');
+      return false;
+    }
+    this.logger.debug(`isActive(): websocket ready state is ${WS_READY_STATES[this.websocket.readyState]}`);
```
[the remark this answers is not in the document]
Comment (filmaj) [reply packages/socket-mode/src/SlackWebSocket.ts:194]: No specific reason! I can change it to an enum if you'd prefer that - lemme know

File: packages/socket-mode/src/SlackWebSocket.spec.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    it('should set a default logger if none provided', () => {
+      new SlackWebSocket({ url: 'https://whatever.com', client: new EventEmitter(), clientPingTimeoutMS: 1, serverPingTimeoutMS: 1 });
+      assert.isTrue(logFactory.called);
+    });
+    it('should not set a default logger if one provided', () => {
+      new SlackWebSocket({ url: 'https://whatever.com', client: new EventEmitter(), clientPingTimeoutMS: 1, serverPingTimeoutMS: 1, logger: new ConsoleLogger() });
+      assert.isFalse(logFactory.called);
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (filmaj) [reply packages/socket-mode/src/SlackWebSocket.spec.ts:29]: The integration tests under the `test/` dir have an actual socket mode server that this module integrates with - is that what you meant or something different?

File: lint-configs/.eslintrc.js [hunk 6]
```diff
@@ -225,7 +225,7 @@ module.exports = {
           },
           {
             selector: 'parameter',
-            format: ['camelCase'],
+            format: ['camelCase', 'snake_case'],
```
[the remark this answers is not in the document]
Comment (filmaj) [reply lint-configs/.eslintrc.js:228]: Some event payload properties that come in from Slack use snake case - I'll leave a comment to explain this.

[the remark this answers is not in the document]
Comment (filmaj) [reply lint-configs/.eslintrc.js:228]: Actually the comment on line 192 explains this already.

File: packages/socket-mode/src/SlackWebSocket.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (WilliamBergamin) [reply packages/socket-mode/src/SlackWebSocket.spec.ts:29]: I didn't see those covered the `ping` pong` cases, but yes this type of test is what I meant

File: packages/socket-mode/src/SlackWebSocket.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamBergamin) [reply packages/socket-mode/src/SlackWebSocket.ts:194]: I prefer the enum
it should make the code self explanatory, the comment on the following will no longer be needed

File: packages/socket-mode/src/SlackWebSocket.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (filmaj) [reply packages/socket-mode/src/SlackWebSocket.spec.ts:29]: Yeah that's fair, I will add more integration tests covering the ping/pong heartbeat stuff.

File: packages/socket-mode/src/SlackWebSocket.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamBergamin) [reply packages/socket-mode/src/SlackWebSocket.ts:194]: Ahh I see, adding the Enum means you need to search for the state based on the integer

In that case sticking to the `array` pattern seems valid

File: packages/socket-mode/README.md [hunk 7]
```diff
@@ -1,29 +1,25 @@
...
-
----
-
 ### Initialize the client
 
-This package is designed to support [**Socket Mode**][socket-mode], which allows your app to receive events from Slack over a WebSocket connection.
-
 The package exports a `SocketModeClient` class. Your app will create an instance of the class for each workspace it communicates with. Creating an instance requires an [**app-level token**][app-token] from Slack. Apps connect to the **Socket Mode** API using an [**app-level token**][app-token], which starts with `xapp`.
 
-Note: **Socket Mode** requires the `connections:write` scope. Navigate to your [app configuration](https://api.slack.com/apps) and go to the **OAuth and Permissions** section to add the scope.
```
Comment (seratch) [inline packages/socket-mode/README.md:26]: Good catch; this had been completely incorrect

**Review by seratch — APPROVED** [review]
Thanks for this great work; I don't see any issues so far

**Comment by filmaj** [comment]
Thanks for the reviews! I did see some CI failures with the integration tests, and I realized some of the sequencing of events / websocket behaviour in the test was not ideal and probably brittle on a weaker machines (like GitHub Actions). Running them once on my local machine tended to pass them, but if I ran the tests like this instead:

```
while npm run test:integration; do echo "OK"; done
```

.. after about 10 or so runs I could trigger a failure.

I made [a change in the tests](https://github.com/slackapi/node-slack-sdk/pull/1781/commits/7b13b097fc8478b91078a48bd9d070b1bc22000c) to make them more resilient (any event-based waiting promises should be created after `start()` completes, not after any timeouts; ensures the event loop has opportunity to execute a few iterations to ensure events get picked up properly by any waiters) and now they pass on both local and CI at a much higher rate.

Also confirming that the longevity test I had set up has been operating flawlessly for a week.

Merging this! Hooray! I will likely publish a 2.0 release of socket-mode soon!

**Pull request merged**